README.md

**CCGMS Term _(6502 assembly source for Commodore 64)_**

**Copyright (C) Craig Smith, 1986-1988, 2016**

*Resurrected and released to world September 5, 2016.
Permission is granted to modify and distribute this work under
the terms of the BSD 3-Clause license.
See LICENSE.txt file for details.*

**Version 5.5+** (ported to build with ca65)

**Version 2016a** (5.5+ with 2400 baud fix, ...)

If you're just looking for a .d64 of the latest image, see the 'dist' folder,
or the 'original' folder for the historical 5.5+.

If you want to build from source or contribute, read on.

This was the last authentic/official version released by me in January 1988.
Over the years others created various hacked versions, both good and bad.
CCGMSTerm 6.01, for example, was a hacked version of (I think) 4.0, pre-
autodialer, and all they did was change my name to "Chris" in the banner
and add a sda decompression wrapper so the disk image was smaller.
CCGMSTerm 7 fixed timing issues with 2400 baud and CCGMS+RLE-8.0 added RLE 
image support.  I do not have corresponding source for any of these
changes, but if people want to submit useful changes, your contributions
are welcome!

The file ccgmsterm.s contains combined source from the original SEQ files
(named 5a.gs,5b.gs,5c.gs and 5d.gs) in CBM Assembler 64 format and chained
together using .FIL directives because the entire source could not
fit into a C64's memory using the CBM editor.  The encoding of this
file has been converted to ASCII (with lowercase labels), and strings
containing the PETSCII left arrow character ($5F) have been changed
to its numeric code so modern or cross-platform assemblers do not
treat it as an underscore.  (The original source files are in the
original folder for posterity.)

This source can be compiled using the ca65 portable 6502 compiler.
Except for the conversion to ASCII and the addition of feature
and conditional directives (.feature/.if/.endif), the format of
this file remains compatible with CBM Assembler 64.
(The .feature directives below enable this compatibility.)

*DEPENDENCIES*

You need ca65 (https://github.com/cc65/cc65.git) to build this,
and the c1541 tool from the VICE emulator (http://vice-emu.sourceforge.net)
to create a .d64 disk image out of it.

*BUILDING*

Just type 'make' to build ccgmsterm, minus the "script kiddie" trivial
encoding scheme that used to protect it from being easily modified with a disk
editor.  You can type "make history" to build a version that still
has this (as well as a baud rate bug on PAL systems that's also fixed
in the normal build; thanks to Jim Drew for discovering this).





CHANGES.md
Version 2016a
-------------
 - Fixed a baud rate bug on PAL systems.  Patch by Jim Drew.
 - Free up some memory by removing gap at location of former Punter buffer
 - Changed version banner to '2016a' unless making historical build
 - Changed author's message.
 - Fixed 2400 baud support using code from "Toward 2400" article
   in Transactor, Vol 9 Issue 3.
 - Fixed XModem upload last character drop bug
 - Add CMD SwiftLink support ($DE00)

Version 5.5+ (09/14/2016)
------------------------
 - Reconstructed 5.5+ changed and added to source

Version 5.5 (09/11/2016)
------------------------
 - Resurrected 5.5 source code and released via github.
 - Source code now cross-assembles under ca65.  Original 5?.gs files replaced
   with single ccgmsterm.s source file.

Version 5.5 (01/15/1988) (historical release by BBS)
------------------------

 - Last official shareware release. (There was actually a 5.5+ released with 
 a minor bug fix, but I've lost the source to it; if I find it, I will update
 this accordingly.)  All other versions (6.01, 8+RLE, EL2, 13.0) out there
 were unoffical hacks, as should probably be obvious if you run them.
 The original CBM assembler source is in the 'original' folder in filenames 
 5a.gs ... 5d.gs.



encode.pl
#!/usr/bin/perl 
#
# This script bootstraps the script-kiddie-protection encoding of an assembled 
# 'original' version of CCGMS Term 5.5 or 5.5+ (built with 'make history'
# target). Alternatively, you could load the assembled ccgmsterm5.5 prg
# on a C64, execute SYS15636 (SYS15631 for 5.5+), and resave it.
#
use strict;
use File::stat;
		              
my $file = shift @ARGV;
my $expectedFileLength = 19279;
my $startEncode = 0x843; # original range of encoded bytes
my $endEncode = 0x3507;
my $decodeSubLength = 74;
my $buf;

die if (!-f $file);

my $sb = stat($file);
my $filesize = $sb->size;

if ($filesize == $expectedFileLength - 5) { # account for change in 5.5+
	$expectedFileLength -= 5;
	$endEncode -= 5;
}

open (F, "<$file") || die;
binmode(F);

die if (read(F,$buf, $filesize) != $filesize);
my $cnt = 0;
my $dbg = 0;
my @bytes = split(//, $buf);
my $noEncode =  ($filesize != $expectedFileLength || ord($bytes[$startEncode]) != 0x78 || ord($bytes[$endEncode-1])!=0xfe);

print STDERR "Binary doesn't match expected form; skipping encoding.\n" if ($noEncode);
foreach(@bytes) {
	if ($dbg) {
		printf ("%02x\n",ord($_))if ($cnt == $startEncode || $cnt == $endEncode-1);
	} elsif (!$noEncode && ($cnt >= $startEncode && $cnt < $endEncode 
		|| $cnt >= $endEncode + $decodeSubLength && $cnt < $filesize)) {
		print(chr(ord($_) ^ 0xaa));
	} else {
		print;
	}
	++$cnt;
}
close(F);


newmodem2400.s
; "newmodem.src"
;
; From "Toward 2400: RS-232 revitied"
; by George Hug
; Transactor, Vol 9, Issue 3
;

; define ccgms to disable features not needed by ccgms term
; (defined when included by ccgmsterm.s)

ribuf   = $f7
robuf   = $f9
baudof  = $0299
ridbe   = $029b
ridbs   = $029c
rodbs   = $029d
rodbe   = $029e
enabl   = $02a1
rstkey  = $fe56
norest  = $fe72
return  = $febc
oldout  = $f1ca
oldchk  = $f21b
ochrin  = $f157
ogetin  = $f13e
findfn  = $f30f
devnum  = $f31f
nofile  = $f701

.ifndef ccgms
* = $ce00 
xx00 jmp rssetup
xx03 jmp inable
xx06 jmp disabl
xx09 jmp rsget
xx0c jmp rsout
     nop
.endif

; start-bit times
strt24 .word   459     ; 2400
strt12 .word   1090    ; 1200
strt03 .word   4915    ;  300
; full-bit times
full24 .word   421     ; 2400
full12 .word   845     ; 1200
full03 .word   3410    ;  300
;-----------------------------------
rssetup lda     #<nmi64
        ldy     #>nmi64
        sta     $0318
        sty     $0319
        lda     #<nchkin
        ldy     #>nchkin
        sta     $031e
        sty     $031f
        lda     #<nbsout
        ldy     #>nbsout
        sta     $0326
        sty     $0327
        lda     #<nchrin
        ldy     #>nchrin
        sta     $0324
        sty     $0325
        lda     #<ngetin
        ldy     #>ngetin
        sta     $032a
        sty     $032b
        rts
;-------------------------
nmi64   pha	; new nmi handler
        txa
        pha
        tya
        pha
nmi128  cld
        ldx     $dd07   ; sample timer B high byte
        lda     #$7f    ; disable CIA NMI's
        sta     $dd0d
        lda     $dd0d   ; read/clear flags
        bpl     notcia  ; (restore key)
        cpx     $dd07   ; timer B timeout since ldx $dd07 above?
        ldy     $dd01   ; (sample pin C)
        bcs     mask    ; no
        ora     #$02    ; tes, set flag in acc.
        ora     $dd0d   ; read/clear flags again
mask    and     enabl   ; mask out non-enabled
        tax             ; these must be serviced
        lsr             ; timer A? (bit 0)
        bcc     ckflag  ; no
        lda     $dd00   ; yes, put bit on pin M
        and     #$fb
        ora     $b5
        sta     $dd00
ckflag  txa
        and     #$10    ; *flag NMI? (bit 4)
        beq     nmion   ; no
strtlo  lda     #$42    ; yes, start-bit to timer B
        sta     $dd06
strthi  lda     #$04    ;
        sta     $dd07
        lda     #$11    ; start timer B counting
        sta     $dd0f
        lda     #$12    ; (flag NMI off, timer B on
        eor     enabl   ; update mask
        sta     enabl
        sta     $dd0d   ; enable new config.
fulllo  lda     #$4d    ; change reload latch
        sta     $dd06   ;  to full-bit time
fullhi  lda     #$03    ;
        sta     $dd07
        lda     #$08    ; # of bits to receive
        sta     $a8
        bne     chktxd  ; branch always
notcia  ldy     #$00
        jmp     rstkey  ; or jmp norest
nmion   lda     enabl   ; re-enable NMI's
        sta     $dd0d
        txa
        and     #$02    ; timer B? (bit 1)
        beq     chktxd  ; no
        tya             ; yes, sample from (sample pin C) above
        lsr
        ror     $aa     ; rs232 is LSB first
        dec     $a8     ; byte finished?
        bne     txd     ; no
        ldy     ridbe   ; yes, byte to buffer
        lda     $aa
        sta     (ribuf),y ; (no over-run test)
        inc     ridbe
        lda     #$00    ; stop timer B
        sta     $dd0f
        lda     #$12    ; timer B NMI off, *flag on
switch  ldy     #$7f    ; disable NMI's
        sty     $dd0d
        sty     $dd0d   ; twice
        eor     enabl   ; update mask
        sta     enabl
        sta     $dd0d   ; enable new config.
txd     txa
        lsr             ; timer A?
chktxd  bcc     exit    ; no
        dec     $b4     ; yes, byte finished?
        bmi     char    ; yes
        lda     #$04    ; no, prep next bit
        ror     $b6     ; (fill with stop bits)
        bcs     store
low     lda     #$00
store   sta     $b5
exit    jmp     return  ; restore regs, rti
char    ldy     rodbs
        cpy     rodbe   ; buffer empty?
        beq     txoff   ; yes
getbuf  lda     (robuf),y ; no, prep next byte
        inc     rodbs
        sta     $b6
        lda     #$09    ; # bits to send
        sta     $b4
        bne     low     ; always - do start bit
txoff   ldx     #$00    ; stop timer A
        stx     $dd0e
        lda     #$01    ; disable timer A NMI
        bne     switch  ; always
;-----------------------------------
disabl  pha             ; turns off modem port
test    lda     enabl
        and     #$03    ; any current activity?
        bne     test    ; yes, test again
        lda     #$10    ; no, disable *flag NMI
        sta     $dd0d
        lda     #$02
        and     enabl   ; currently receiving?
        bne     test    ; yes, start over
        sta     enabl   ; all off, update mask
        pla
        rts
;-----------------------------------
nbsout  pha             ; new bsout
        lda     $9a
        cmp     #$02
        bne     notmod
        pla
rsout   sta     $9e     ; output to modem
        sty     $97
point   ldy     rodbe
        lda     $9e
        sta     (robuf),y ; not official 'til sty rodbe below
        iny
        cpy     rodbs   ; buffer full?
        beq     fulbuf  ; yes
        sty     rodbe   ; no, bump pointer
strtup  lda     enabl
        and     #$01    ; transmitting now?
        bne     ret3    ; yes
        sta     $b5     ; no, prep start bit
        lda     #$09
        sta     $b4     ;  # bits to send
        ldy     rodbs
        lda     (robuf),y
        sta     $b6     ;  and next byte
        inc     rodbs
        lda     baudof  ; full TX bit time to timer A
        sta     $dd04
        lda     baudof+1
        sta     $dd05
        lda     #$11    ; start timer A
        sta     $dd0e
        lda     #$81    ; enable timer A NMI
change  sta     $dd0d   ; NMI clears flag if set
        php             ; save IRQ status.
        sei             ; disable IRQ's
        ldy     #$7f    ; disable NMI's
        sty     $dd0d
        sty     $dd0d   ; twice
        ora     enabl   ; update mask
        sta     enabl
        sta     $dd0d   ; enable new config.
        plp             ; restore IRQ status
ret3    clc
        ldy     $97
        lda     $9e
        rts
fulbuf  jsr     strtup
        jmp     point
notmod  pla             ; back to old bsout.
        jmp     oldout
;-----------------------------------
nchkin
.ifdef ccgms ; if using swiftlib, allow chkin of modem when swiftlink active
.if swiftlib ; to 'work'. Overridden chrin/getin/chrout routines check this
; to map to slGetByte/slPutByte calls
        cpx     #modemln
        bne     nchkin2
        lda     swiftout
        beq     nchkin2
        lda     #$02
        sta     $99
        rts
nchkin2
.endif
.endif
        jsr     findfn  ; new chkin
        bne     nosuch
        jsr     devnum
        lda     $ba
        cmp     #$02
        bne     back
        sta     $99
inable  sta     $9e     ; enable rs232 input
        sty     $97
bauds   lda     baudof+1 ; set receive to same
        and     #$06     ;  baud rate as xmit
        tay
        lda     strt24,y
        sta     strtlo+1 ; overwrite code setting timer B (self mod)
        lda     strt24+1,y
        sta     strthi+1
        lda     full24,y
        sta     fulllo+1
        lda     full24+1,y
        sta     fullhi+1
        lda     enabl
        and     #$12    ; flag or timer B on?
        bne     ret1    ; yes
        sta     $dd0f   ; no, stop timer B
        lda     #$90    ; turn on flag NMI
        jmp     change
nosuch  jmp     nofile
back    lda     $ba
        jmp     oldchk
;-----------------------------------
nchrin  lda     $99     ; new chrin
        cmp     #$02
        beq     rsget
        jmp     ochrin
ngetin  ldx     $99     ; new getin
        lda     #$00
        cpx     #$02
        beq     rsget
        jmp     ogetin
;
rsget   sta     $9e     ; input from modem
        sty     $97
.ifdef ccgms ; see comment at nchkin
.if swiftlib
        ldy     swiftout
        beq     rsget1
        jsr     slGetByte
        ldy     $97
        rts
rsget1
.endif
.endif
        ldy     ridbs
        cpy     ridbe   ; buffer empty?
        beq     ret2    ; yes.
        lda     (ribuf),y ; no, fetch character
        sta     $9e
        inc     ridbs
ret1    clc             ; cc = char in acc.
ret2    ldy     $97
        lda     $9e
last    rts             ; vc = buffer was empty



swiftlib.s

;This file has been modified slightly to be compatible
;with the ca65 assembler for use by CCGMS Term.
;Other than minor asm syntax changes and commenting out of .obj
;directives,, very few modifications have been made to the 
;original swiftlib. Those modifications are guarded
;with .ifdef ccgms.
.feature labels_without_colons 

;Original file header comments follow:

;SwiftLink/Turbo-232 v0.91 device driver, by Craig Bruce, 21-Aug-2000.

;This software is Public Domain.  It is in Buddy assembler format.

;This device driver uses the SwiftLink RS-232 Serial Cartridge, available from
;Creative Micro Designs, Inc, and also supports the extensions of the Turbo232
;Serial Cartridge.  Both devices are based on the 6551 ACIA chip.  It also
;supports the "hacked" SwiftLink with a 1.8432 MHz crystal.

;The code assumes that the kernal + I/O are in context.  On the C128, call
;it from Bank 15.  On the C64, don't flip out the Kernal unless a suitable
;NMI catcher is put into the RAM under then Kernal.  For the SuperCPU, the
;interrupt handling assumes that the 65816 is in 6502-emulation mode.

useC128 = 0    ;1=C128, 0=C64

.if useC128
.ifndef ccgms
   .org $1300
.endif
   ;.obj "@0:swiftlib128"
   slHead     = $c8  ;(1) ;buffer head pointer
   slTail     = $c9  ;(1) ;buffer tail pointer
   slFreeCnt  = $ca  ;(1) ;bytes free in buffer
   slRtsOff   = $cb  ;(1) ;must be non-zero (DTR,ints active)
   slErrors   = $b4  ;(1) ;number of bytes received in error, low byte
   slWork     = $b5  ;(2) ;temp zp work area
   slNmiExit  = $ff33     ;exit address for nmi
   useC64     = 0
.else
.ifndef ccgms
   .org $c000
.endif
   ;.obj "@0:swiftlib64"
   slHead     = $f7  ;(1) ;buffer head pointer
   slTail     = $f8  ;(1) ;buffer tail pointer
   slFreeCnt  = $f9  ;(1) ;bytes free in buffer
   slRtsOff   = $fa  ;(1) ;must be non-zero (DTR,ints active)
   slErrors   = $a7  ;(1) ;number of bytes received in error, low byte
   slWork     = $b4  ;(2) ;temp zp work area
   slNmiExit  = $febc     ;exit address for nmi
   useC64     = 1
.endif

;*----------------------------------------------------------------------------*
;  user interface jump table and variables
;*----------------------------------------------------------------------------*

slJumpTable = *
   jmp slInit      ;+0  ;( .AY=iobase, .X=hackedSlFlag )
   jmp slParms     ;+3  ;( .A=params, .X=parity ) : .CS=err#.A
   jmp slShutdown  ;+6  ;()
   jmp slGetByte   ;+9  ;() : .A=byte, .X=notEmpty, .CS=err#.A
   jmp slPutByte   ;+12 ;( .A=byte ) : .CS=err#.A
   jmp slPause     ;+15 ;()
   jmp slUnpause   ;+18 ;()
   jmp slStatus    ;+21 ;() : .A=status, .X=errorCount

slDropCnt     .byte 0,0,0,0    ;(4) ;number of bytes lost from rx buffer full

;*----------------------------------------------------------------------------*
;  global variables
;*----------------------------------------------------------------------------*

slInitialized .byte $00 ;(1) ;flag indicating driver is initialized
slStopped     .byte 0   ;(1) ;flow-stopped flag
slIoAddr      .word 0    ;(2) ;base address of swiftlink registers
slIoPtr       = slWork  ;(2) ;pointer to i/o page
slTemp        .byte 0,0,0,0    ;(4) ;temporary storage
slTurbo232    .byte 0    ;(1) ;flag indicating turbo-232
slHackedFlag  .byte 0    ;(1) ;flag indicating hacked-crystal swiftlink
slCpuSpeed    .byte 0    ;(1) ;in MHz
slSendHead    .byte 0    ;(1) ;head of send buffer
slSendTail    .byte 0    ;(1) ;tail of send buffer
slSendFreeCnt .byte 0    ;(1) ;number of bytes free in send buffer
slNmiContinue .byte $4c  ;(1) ;JMP instruction for NMI save -- continue
slNmiSave     .byte 0,0  ;(2) ;normal NMI handler
slTryHard     .byte 0    ;(1) ;how hard to try to send from send buffer
slBaudCode    .byte 0    ;(1) ;current baud in effect

slRegData     = 0       ;offset of swiftlink data register
slRegStatus   = 1       ;offset of swiftlink status register
slRegCommand  = 2       ;offset of swiftlink command register
slRegControl  = 3       ;offset of swiftlink control register
slRegClock    = 7       ;offset of turbo232 external baud-rate generator

slErrNotInitialized = $01
slErrBaudTooFast    = $02
slErrBaudNotAvail   = $03
slErrNoData         = $04
slErrOverflow       = $05

;*----------------------------------------------------------------------------*
; slInit - swiftlink initialize.
;*----------------------------------------------------------------------------*

slInit = *  ;( .AY=iobase, .X=hackedSLflag )
   ;** shut down if started
   bit slInitialized
   bpl :+
   pha
   txa
   pha
   tya
   pha
   jsr slShutdown
   pla
   tay
   pla
   tax
   pla

   ;** set io page, hacked-crystal
:  sta slIoAddr+0
   sty slIoAddr+1
   sta slIoPtr+0
   sty slIoPtr+1
   txa
   beq :+
   lda #$ff
:  sta slHackedFlag

   ;** check for turbo-232
   lda #$00
   ldy #slRegControl
   sta (slIoPtr),y
   ldy #slRegClock
   ldx #$00
   lda (slIoPtr),y
   beq :+
   dex
:  stx slTurbo232

   ;** get C128/C64 cpu speed
.if useC128
   lda $d030
   and #$01
   clc
   adc #1
   sta slCpuSpeed
.else
   lda #1
   sta slCpuSpeed
.endif

   ;** check for super-cpu at 20 MHz
   bit $d0bc
   bmi :+
   bit $d0b8
   bvs :+
   lda #20
   sta slCpuSpeed
   
   ;** initialize buffers & control
:  lda #0
   sta slHead
   sta slSendHead
   sta slTail
   sta slSendTail
   sta slErrors
   sta slStopped
   lda #255
   sta slFreeCnt
   sta slSendFreeCnt

   ;** set up nmi's
   lda $318
   ldy $319
   sta slNmiSave+0
   sty slNmiSave+1
   lda #<slNmiHandler
   ldy #>slNmiHandler
   sta $318
   sty $319
   ldx slIoAddr+0
   ldy slIoAddr+1
   txa
   clc
   adc slNmiIoRef1+0
   sta slNmiIoRef1+0
   sty slNmiIoRef1+1
   txa
   clc
   adc slNmiIoRef2+0
   sta slNmiIoRef2+0
   sty slNmiIoRef2+1
   txa
   clc
   adc slNmiIoRef3+0
   sta slNmiIoRef3+0
   sty slNmiIoRef3+1
   txa
   clc
   adc slNmiIoRef4+0
   sta slNmiIoRef4+0
   sty slNmiIoRef4+1

   ;** set default to 2400-8N1, enable interrupts
   ldy #slRegData
   lda (slIoPtr),y
   ldy #slRegStatus
   lda (slIoPtr),y
   ldy #slRegControl
   lda #$18
   bit slHackedFlag
   bpl :+
   lda #$1a
:  sta (slIoPtr),y
   ldy #slRegCommand
   lda #$01
   sta slRtsOff
   ora #$08
   sta (slIoPtr),y
   lda #$06
   sta slBaudCode

   ;** return
   lda #$ff
   sta slInitialized
   clc
   rts

;*----------------------------------------------------------------------------*
; slParms - swiftlink set communication parameters.
;
; baud rates              stops     word    |   parity
; ---------------------   -----     -----   |   ---------
; $00=50     $08=9600     $00=1     $00=8   |   $00=none
; $01=110    $09=19200    $80=2     $20=7   |   $20=odd
; $02=134.5  $0a=38400              $40=6   |   $60=even
; $03=300    $0b=57600              $60=5   |   $A0=mark
; $04=600    $0c=115200                     |   $E0=space
; $05=1200   $0d=230400
; $06=2400   $0e=future
; $07=4800   $0f=future
;*----------------------------------------------------------------------------*

slParms = *  ;( .A=params, .X=parity ) : .CS=err#.A
   ;** check initialized
   bit slInitialized
   bmi :+
   lda #slErrNotInitialized
   sec
   rts

   ;** save new parity
:  sta slTemp
   lda slIoAddr+0
   ldy slIoAddr+1
   sta slIoPtr+0
   sty slIoPtr+1
   txa
   and #%11100000
   ora #%00000001
   sta slTemp+1

   ;** check cpu speed against baud rate
   lda slTemp
   and #$0f
   cmp #$0c
   bcc :++
   ldx slCpuSpeed
   cpx #1+1
   bcc :+
   cmp #$0c
   beq :++
   cpx #4
   bcs :++
:  sec
   lda #slErrBaudTooFast
   rts
   
   ;** set baud/parameters
:  lda slTemp
   and #$0f
   tax
   lda slNormBauds,x
   bit slHackedFlag
   bpl :+
   lda slHackBauds,x
:  cmp #$ff
   bne :+
   sec
   lda #slErrBaudNotAvail
   rts
:  tax
   and #$30
   beq :+
   bit slTurbo232
   bmi :+
   sec
   lda #slErrBaudNotAvail
   rts
:  lda slTemp
   and #$0f
   sta slBaudCode
   lda slTemp
   and #%11100000
   ora #%00010000
   sta slTemp
   txa
   and #$0f
   ora slTemp
   ldy #slRegControl
   sta (slIoPtr),y
   txa
   and #$30
   beq :+
   lsr
   lsr
   lsr
   lsr
   eor #$03
   ldy #slRegClock
   sta (slIoPtr),y

   ;** set new parity
:  lda slTemp+1
   sta slRtsOff
   ora #$08
   ldy #slRegCommand
   sta (slIoPtr),y
   clc
   rts

slNormBauds = *
   .byte $ff,$ff,$ff,$05,$06,$07,$08,$0a,$0c,$0e,$0f,$10,$20,$30,$ff,$ff
slHackBauds = *
   .byte $01,$03,$04,$06,$07,$08,$0a,$0c,$0e,$0f,$ff,$ff,$00,$ff,$ff,$ff
     ;in:  0   1   2   3   4   5   6   7   8   9   a   b   c   d   e   f
     ;baud50 110 134   3   6  12  24  48  96  19  38  57 115 230 exp exp
     ;out masks: $0F=slBaud, val$FF=err
     ;           $30=t232ExtBaud: $00=none, $10=57.6, $20=115.2, $30=230.4

;*----------------------------------------------------------------------------*
; slShutdown - shut down swiftlink.
;*----------------------------------------------------------------------------*

slShutdown = * ;()
   ;** check initialized
   bit slInitialized
   bmi :+
   clc
   rts

   ;** stop interrupts, drop DTR
:  lda slIoAddr+0
   ldy slIoAddr+1
   sta slIoPtr+0
   sty slIoPtr+1
   lda slRtsOff
   and #%11100010
   ora #%00000010
   ldy #slRegCommand
   sta (slIoPtr),y

   ;** restore NMI vector
   lda slNmiSave+0
   ldy slNmiSave+1
   sta $318
   sty $319

   ;** flag uninitialized
   lda #$00
   sta slInitialized
   rts

;*----------------------------------------------------------------------------*
; slGetByte - get byte from receive buffer
;   arguments:
;     none
;   return values:
;     .A = byte value retrieved from buffer
;     .X = zero if no data available, otherwise non-zero.  more specifically,
;          it returns the number of bytes available in the buffer, including
;          the byte being returned
;     .C = clear for okay, set on error with the error code in .A
;*----------------------------------------------------------------------------*

slGetByte = *  ;() : .A=byte, .X=notEmpty, .CS=err#.A
   ;** check initialized
   jsr slCheckInitialized
   bcc :+
   rts

   ;** check for bytes to send
:  ldx slSendFreeCnt
   cpx #$ff
   beq :+
   lda #$00
   jsr slTryToSend

   ;** check for buffer empty
:  lda slFreeCnt
   cmp #$ff
   bne :+
   ldx #0
   txa
   clc
   rts

   ;** check for flow stopped & enough free: release flow control
:  ldx slStopped
   beq :+
   cmp #63
   bcc :+
   lda #$00
   sta slStopped
   lda slRtsOff
   ora #%00001000
   ldy #slRegCommand
   sta (slIoPtr),y

   ;** get byte from buffer
:  ldx slHead
   ldy slReceiveBuf,x
   inc slHead
   sec
   lda #$ff
   sbc slFreeCnt  ;race condition--count may be conservative
   inc slFreeCnt
   tax
   tya
   clc
   rts

;*----------------------------------------------------------------------------*
; slPutByte - put byte to swiftlink
;   arguments:
;     .A = byte value to send
;   return values:
;     .C = clear for okay, set on error with the error code in .A
;*----------------------------------------------------------------------------*

slPutByte = *  ;( .A=byte ) : .CS=overflow
   ;** check initialized
   jsr slCheckInitialized
   bcc :+
   rts

   ;** try to send
:  ldx slSendFreeCnt
   cpx #$ff
   beq :+
   pha
   lda #$00
   jsr slTryToSend
   pla

   ;** check for send-buffer overflow
:  ldx slSendFreeCnt
   bne :+
   tax
   lda #slErrOverflow
   sec
   rts

   ;** put byte into send buffer & send
:  ldx slSendTail
   sta slSendBuf,x
   inc slSendTail
   dec slSendFreeCnt
   lda #$ff
   jsr slTryToSend
   clc
   rts

slTryToSend = *  ;( .A=tryHard )  ;assumes slIoPtr is set
   ;** check for something to send
   sta slTryHard
   lda slSendFreeCnt
   cmp #$ff
   bne :+
   rts

   ;** check for flow stopped
:  lda slStopped
   beq :+
   rts

   ;** check that swiftlink is ready to send
:  ldy #slRegStatus
   lda (slIoPtr),y
   and #$10
   bne :+
   lda slTryHard      ;keep trying if must try hard
   bmi slTryToSend
   rts

   ;** send byte and try again
:  ldx slSendHead
   lda slSendBuf,x
   ldy #slRegData
   sta (slIoPtr),y
   inc slSendHead
   inc slSendFreeCnt
   jmp slTryToSend

;*----------------------------------------------------------------------------*
; slPause - assert flow control & disable swiftlink interrupts
;   arguments:
;     none
;   return values:
;     .C = clear for okay, set on error with the error code in .A
;*----------------------------------------------------------------------------*

slPause = *  ;()
   ;** check initialized
   jsr slCheckInitialized
   bcc :+
   rts

   ;** assert flow control
:  lda slRtsOff
   sta slStopped
   ldy #slRegCommand
   sta (slIoPtr),y

   ;** delay for flow stop to be received
   ldx slBaudCode
   lda slPauseTimes,x
   jsr slDelayMs

   ;** stop rx interrupts
   lda slRtsOff
   ora #$02
   ldy #slRegCommand
   sta (slIoPtr),y
   clc
   rts

   ;** delay times: 32 byte-receive times in milliseconds, or 100 max
   ;** formula = 320,000 / baud
   slPauseTimes = *
   .byte 100,100,100,100,100,100,100,067,034,017,009,006,003,002,001,001
     ;in:  0   1   2   3   4   5   6   7   8   9   a   b   c   d   e   f
     ;baud50 110 134   3   6  12  24  48  96  19  38  57 115 230 exp exp

;*----------------------------------------------------------------------------*
; slUnpause - re-enable swiftlink interrupts & release flow control
;   arguments:
;     none
;   return values:
;     none
;*----------------------------------------------------------------------------*

slUnpause = *  ;()
   ;** check initialized
   jsr slCheckInitialized
   bcc :+
   rts

   ;** re-enable rx interrupts & release flow control
:  lda #$00
   sta slStopped
   lda slRtsOff
   ora #%00001000
   ldy #slRegCommand
   sta (slIoPtr),y

   ;** poll for stalled char & exit
   jsr slPollReceive
   clc
   rts

;*----------------------------------------------------------------------------*
; slStatus - get status of swiftlink driver
;   arguments:
;   return values:
;*----------------------------------------------------------------------------*

slStatus = *  ;( ) : .A=status, .X=errorCounter, .CS=err#.A
   ;** check initialized
   jsr slCheckInitialized
   bcc :+
   rts

   ;** get status
:  ldy #slRegStatus
   lda (slIoPtr),y
   pha
   jsr slPollReceive  ;bug-recovery hack
   pla
   ldx slErrors
   clc
   rts

;*----------------------------------------------------------------------------*
; NMI handler
; C128 NMI overhead=76 cycles: int=7, maxLatency=6, ROMenter=33, ROMexit=30
; C64  NMI overhead=76 cycles: int=7, maxLatency=6, ROMenter=34, ROMexit=29
;
; timing: normal=76+43+9=128 cycles, assertFlow=76+52+9=137 cycles
;
; C128 @ 115.2k: 177 cycles avail (fast)
; C64  @  57.6k: 177 cycles avail, worstAvail=177-43? = 134
; SCPU @ 230.4k: 868 cycles avail: for a joke!
;*----------------------------------------------------------------------------*

slNmiHandler = *
.if useC64
   pha
.endif
   slNmiIoRef1 = *+1
   lda $de00+1        ;(4) ;status ;check for byte received
   and #$08           ;(2)
   beq slNmiNorm      ;(2*)
.if useC64
   cld
   txa
   pha
   tya
   pha
.endif
   slNmiIoRef2 = *+1
   lda $de00+1        ;(4) opt ;status ;check for receive errors
   and #$07           ;(2) opt
   beq :+              ;(3*)opt
   inc slErrors       ;(5^)opt
   slNmiIoRef3 = *+1
:  lda $de00+0        ;(4) ;data  ;get byte and put into receive buffer
   ldy slTail         ;(3)
   ldx slFreeCnt      ;(3)
   beq slDrop         ;(2*)
   sta slReceiveBuf,y ;(5)
   inc slTail         ;(5)
   dec slFreeCnt      ;(5)
   cpx #33            ;(2)  ;check for buffer space low
   bcc :+              ;(2*)
   jmp slNmiExit      ;(3)
:  lda slRtsOff       ;(3)  ;assert flow control if buffer space too low
   slNmiIoRef4 = *+1
   sta $de00+2        ;(4) ;command
   sta slStopped      ;(3)
   jmp slNmiExit      ;(3)

slDrop = *
   inc slDropCnt+0    ;not time-critical
   bne :+
   inc slDropCnt+1
   bne :+
   inc slDropCnt+2
   bne :+
   inc slDropCnt+3
:  jmp slNmiExit

slNmiNorm = *
.if useC64
   pla
.endif
   jmp slNmiContinue

;*----------------------------------------------------------------------------*
; slCheckInitialized  -  internal check if initialized & set slIoPtr
;   doesn't change any registers
;*----------------------------------------------------------------------------*

slCheckInitialized = *
   ;** check initialized
   bit slInitialized
   bmi :+
   lda #slErrNotInitialized
   sec
   rts

   ;** set ioptr
:  pha
   lda slIoAddr+0
   sta slIoPtr+0
   lda slIoAddr+1
   sta slIoPtr+1
   pla
   clc
   rts

;*----------------------------------------------------------------------------*
; slPollReceive - poll for rx char
;   assumes that slIoPtr is set
;   This function is useful in odd cases where the 6551 has a character in
;   it but it fails to raise an NMI.  It might be edge-triggering conditions?
;   Actually, I'm not entirely sure that this condition can still arise, but
;   calling this function does no harm.
;*----------------------------------------------------------------------------*

slPollReceive = *
   ldy #slRegStatus
   lda #$08
   and (slIoPtr),y
   beq :+
   and (slIoPtr),y
   beq :+
   ldy #slRegData
   lda (slIoPtr),y
   ldx slFreeCnt
   beq :+
   ldx slTail
   sta slReceiveBuf,x
   inc slTail
   dec slFreeCnt
:  rts

;*----------------------------------------------------------------------------*
;  slDelayMs : delay for given number of milliseconds
;    This implementation isn't very rigerous; it merely delays for the
;    approximate number of clock cycles for the processor speed.
;    Algorithm:
;       repeat for number of milliseconds: 
;          repeat for number of MHz of cpu speed:
;             delay for 1017 clock cycles
;*----------------------------------------------------------------------------*

slDelayMs = *  ;( .A=milliseconds )
:  ldy slCpuSpeed
:  ldx #203   ;(2)
:  dex        ;(2)
   bne :-      ;(3) // 1017 cycles
   dey
   bne :--
   sec
   sbc #1
   bne :---
   rts

;*----------------------------------------------------------------------------*
;  send and receive buffers: 256 bytes each
;*----------------------------------------------------------------------------*

.ifdef ccgms
slEnd = * ; ccgms instead will define the send and receive buf to $ce00/$cf00
; in the same place it used for the kernal rs232 buffers
.else
slReceiveBuf = *
.res 256
slSendBuf = slReceiveBuf+256
.res 256
slEnd = slSendBuf+256
.endif
